### PR TITLE
Adds Jupyterlab instructions to readme, refactors scripts to use _pretty_print_failure, adds code tree

### PR DESCRIPTION
**Pull request by jjgmckenzie**

### CODE REVIEW

**Comment by jjgmckenzie** [comment]
Note: automatic jupyterlab on boot is not added in this PR as requested in the task, I had a few too many questions about the correct approach (The most obvious solution, using systemd, breaks the barrier we are trying to enforce between userspace and kernelspace) and did not want to delay further.

File: jupyterlab/start-jupyterlab.sh [hunk 1]
```diff
@@ -3,20 +3,17 @@
 # toby@indieagi.org
 
 # Source common functions that start with _
-source $JUPYTERLAB_SCRIPT_DIR/common-functions.sh
+source $JUPYTERLAB_SCRIPT_DIR/../common-functions.sh
 # Fixed name for the virtual environment
 VENV_NAME="jupyterlab-venv"
 # Check if the directory already exists
 if [ ! -d "$JUPYTERLAB_SCRIPT_DIR/$VENV_NAME" ]; then
-  _pretty_print "Virtual environment does not exist; creating"
```
Comment (tobkin) [inline jupyterlab/start-jupyterlab.sh:11]: refactor to, and create, _pretty_print_info

File: jupyterlab/symlink-jupyterlab.sh [hunk 2]
```diff
@@ -1,13 +1,19 @@
...
+source ../common-functions.sh
 
 # Check if ~/.zshrc exists; error if not
 if [ ! -f ~/.zshrc ]; then
-  _pretty_print "zsh is not installed! install zsh before following these instructions."
+  _pretty_print_failure "zsh is not installed! install zsh before following these instructions."
   exit 0
 fi
 
+# Check if the current shell is "zsh"
```
Comment (tobkin) [inline jupyterlab/symlink-jupyterlab.sh:11]: add to start-jupyterlab.sh as well

Comment (jjgmckenzie) [reply jupyterlab/symlink-jupyterlab.sh:11]: :+1:

File: jupyterlab/start-jupyterlab.sh (hunk 1 above)
Comment (jjgmckenzie) [reply jupyterlab/start-jupyterlab.sh:11]: :+1:

**Pull request merged**